Implement `:bulk/create` action for SQL JDBC. Tons of bug fixes

camsaul opened the pull request:
Implements or fixes

- #23848
- #23927
- #23930
- #23931
- #23932
- #23933
- #23934
- #23936

This also adds support for `NOT NULL` to test datasets and makes `test-data.categories.name` `NOT NULL` which makes `INSERT` much easier to test. I added tests around `POST /api/action/row/create` with bad data (previously we had no such tests). I also spent a while refactoring `metabase.api.action-test` a bit to make things clearer more explicit.

camsaul committed:
Actions multimethod refactor.

camsaul committed:
Merge branch 'pod-writeback' into actions-multimethod-refactor

camsaul committed:
Merge branch 'pod-writeback' into actions-multimethod-refactor

camsaul committed:
Fix function called with wrong number of args

camsaul committed:
Test fixes

camsaul committed:
Bulk create action (second pass)

camsaul committed:
Merge branch 'pod-writeback' into actions-multimethod-refactor

camsaul committed:
Appease clj-kondo

camsaul committed:
Merge branch 'actions-multimethod-refactor' into bulk-insert-action

camsaul committed:
Fix bulk insert happy path

camsaul committed:
Implement/fix everything

camsaul committed:
Merge branch 'pod-writeback' into bulk-insert-action

camsaul commented on .clj-kondo/config.edn:
```diff
@@ -345,45 +345,49 @@
+ {metabase-enterprise.serialization.test-util/with-random-dump-dir      clojure.core/let
+  metabase.api.common/let-404                                           clojure.core/let
+  metabase.db.data-migrations/defmigration                              clojure.core/def
+  metabase.db.liquibase/with-liquibase                                  clojure.core/let
+  metabase.driver.mongo.query-processor/mongo-let                       clojure.core/let
+  metabase.driver.mongo.util/with-mongo-connection                      clojure.core/let
+  metabase.driver.sql-jdbc.actions/with-jdbc-connection                 clojure.core/let
+  metabase.driver.sql-jdbc.execute.diagnostic/capturing-diagnostic-info clojure.core/fn
+  metabase.integrations.ldap/with-ldap-connection                       clojure.core/fn
+  metabase.mbql.schema.macros/defclause                                 clj-kondo.lint-as/def-catch-all
+  metabase.models.setting.multi-setting/define-multi-setting            clojure.core/def
+  metabase.models.setting/defsetting                                    clj-kondo.lint-as/def-catch-all
+  metabase.public-settings.premium-features/defenterprise               clj-kondo.lint-as/def-catch-all
+  metabase.public-settings.premium-features/defenterprise-schema        clj-kondo.lint-as/def-catch-all
+  metabase.public-settings.premium-features/define-premium-feature      clojure.core/def
+  metabase.query-processor.error-type/deferror                          clojure.core/def
+  metabase.sync.util/sum-for                                            clojure.core/for
+  metabase.sync.util/with-emoji-progress-bar                            clojure.core/let
+  metabase.test.data.interface/defdataset                               clojure.core/def
+  metabase.test.data.interface/defdataset-edn                           clojure.core/def
+  metabase.test/are+                                                    clojure.test/are
+  metabase.test/defdataset                                              clojure.core/def
+  metabase.test/with-open-channels                                      clojure.core/let
+  metabase.test/with-temp-file                                          clojure.core/let
+  metabase.test/with-user-in-groups                                     clojure.core/let
+  metabase.util.files/with-open-path-to-resource                        clojure.core/let
+  metabase.util.ssh/with-ssh-tunnel                                     clojure.core/let
+
+  clojure.core.logic/defne           clj-kondo.lint-as/def-catch-all
+  clojurewerkz.quartzite.jobs/defjob clojure.core/defn
+  honeysql.helpers/defhelper         clj-kondo.lint-as/def-catch-all
+  honeysql.util/defalias             clojure.core/def
+  monger.operators/defoperator       clojure.core/def
+  potemkin.types/defprotocol+        clojure.core/defprotocol
+  potemkin.types/defrecord+          clojure.core/defrecord
+  potemkin.types/deftype+            clojure.core/deftype
+  potemkin/defprotocol+              clojure.core/defprotocol
+  potemkin/defrecord+                clojure.core/defrecord
+  potemkin/deftype+                  clojure.core/deftype
+  toucan.db/with-call-counting       clojure.core/fn}
```
I added `metabase.driver.sql-jdbc.actions/with-jdbc-connection clojure.core/let` here. It really made me unhappy that these were unsorted and we seemed to have two groups -- Metabase stuff and library stuff -- but `toucan.db/with-call-counting` was grouped with the Metabase stuff. 

So I sorted these and fixed the groups and nicely indented them as well.

camsaul commented on src/metabase/actions.clj:
```diff
@@ -37,7 +38,7 @@
 (defmulti normalize-action-arg-map
   "Normalize the `arg-map` passed to [[perform-action!]] for a specific `action`."
-  {:arglists '([action arg-map])}
+  {:arglists '([action arg-map]), :added "0.44.0"}
```
Forgot to add this stuff earlier. But all driver methods should have `^:added` metadata so we know what version we introduced them in.

I know this stuff probably won't actually ship as part of 44, but we can go update these numbers down the road if we need to -- so 44 is mostly a placeholder.

camsaul commented on src/metabase/actions.clj:
```diff
@@ -114,15 +115,15 @@
     ;; Check that Actions are enabled for this specific Database.
     (let [{database-id :database}                         arg-map
-          {db-settings :settings, driver :engine, :as db} (Database database-id)]
+          {db-settings :settings, driver :engine, :as db} (api/check-404 (Database database-id))]
```
Improved error handling

camsaul commented on src/metabase/actions.clj:
```diff
@@ -145,39 +146,57 @@
+;;; Anything else required depends on the action type.
 
 (s/def :actions.args/id
   (s/and integer? pos?))
 
-(s/def :actions.args.crud.common/database
+(s/def :actions.args.common/database
```
I refactored this stuff a little bit more. `:database` is now a required key for all actions that use `perform-action!`. `{:query {:source-table <id>}}` is not since it's only used by the existing `:row/*` actions.

camsaul replied:
I tweaked the spec names a bit to reflect this

camsaul commented on src/metabase/actions.clj:
```diff
@@ -145,39 +146,57 @@
 
 ;;; the various `:row/*` Actions all treat their args map as an MBQL query.
 
-(defn- normalize-as-mbql-query [query]
-  (let [query (assoc (mbql.normalize/normalize query)
-                     :type :query)]
-    (try
-      (schema/validate mbql.s/Query query)
-      (catch Exception e
-        (throw (ex-info
-                (ex-message e)
-                {:exception-data (ex-data e)
-                 :status-code    400}))))
-    query))
+(defn- normalize-as-mbql-query
+  "Normalize `query` as an MBQL query. Optional arg `:exclude` is a set of *normalized* keys to exclude from recursive
+  normalization, e.g. `:create-row` for the `:row/create` Action (we don't want to normalize the row input since
+  preserving case and `snake_keys` in the request body is important)."
+  ([query]
+   (let [query (mbql.normalize/normalize (assoc query :type :query))]
+     (try
+       (schema/validate mbql.s/Query query)
+       (catch Exception e
+         (throw (ex-info
+                 (ex-message e)
+                 {:exception-data (ex-data e)
+                  :status-code    400}))))
+     query))
+
+  ([query & {:keys [exclude]}]
+   (let [query (update-keys query mbql.u/normalize-token)]
+     (merge (select-keys query exclude)
+            (normalize-as-mbql-query (apply dissoc query exclude))))))
```
Fix for #23930 -- we can now exclude keys from normalization e.g. `:create-row` for the `:row/create` action.

camsaul commented on src/metabase/api/action.clj:
```diff
@@ -23,7 +24,9 @@
   [action-namespace action-name table-id :as {:keys [body]}]
   (let [action (keyword action-namespace action-name)]
-    (actions/perform-action! action {:table-id table-id, :arg body})))
+    (actions/perform-action! action {:database (api/check-404 (table/table-id->database-id table-id))
```
This was needed since `:database` is actually a required key in `:actions.args/common`. (I added this endpoint earlier today in #23922, which I spun off from this PR, but it was unused; this PR is the first thing to actually use it.)

camsaul commented on src/metabase/driver/h2/actions.clj:
```diff
@@ -0,0 +1,22 @@
+  "Method impls for [[metabase.driver.sql-jdbc.actions]] for `:h2`."
+  (:require [metabase.driver.sql-jdbc.actions :as sql-jdbc.actions]))
+
+(defmethod sql-jdbc.actions/base-type->sql-type :h2
+  [_driver]
+  {:type/BigInteger     "BIGINT"
+   :type/Boolean        "BOOL"
+   :type/Date           "DATE"
+   :type/DateTime       "DATETIME"
+   :type/DateTimeWithTZ "TIMESTAMP WITH TIME ZONE"
+   :type/Decimal        "DECIMAL"
+   :type/Float          "FLOAT"
+   :type/Integer        "INTEGER"
+   :type/Text           "VARCHAR"
+   :type/Time           "TIME"})
```
Part of the fix for #23933

camsaul commented on src/metabase/driver/postgres.clj:
```diff
@@ -14,6 +14,7 @@
             [metabase.driver.common :as driver.common]
             [metabase.driver.ddl.interface :as ddl.i]
             [metabase.driver.ddl.postgres :as ddl.postgres]
+            [metabase.driver.postgres.actions :as postgres.actions]
```
Part of the fix for https://github.com/metabase/metabase/issues/23933

camsaul commented on src/metabase/driver/postgres/actions.clj:
```diff
@@ -0,0 +1,81 @@
+(ns metabase.driver.postgres.actions
```
I moved all the Postgres-specific stuff here as part of the fix for part of the fix for #23933

camsaul commented on src/metabase/driver/postgres/actions.clj:
```diff
@@ -0,0 +1,81 @@
+
+;; For Postgres creating a Savepoint and rolling it back on error seems to be enough to let the parent transaction
+;; proceed if some particular statement encounters an error.
+(defmethod sql-jdbc.actions/do-nested-transaction :postgres
+  [_driver ^java.sql.Connection conn thunk]
+  (let [savepoint (.setSavepoint conn)]
+    (try
+      (thunk)
+      (catch Throwable e
+        (.rollback conn savepoint)
+        (throw e))
+      (finally
+        (.releaseSavepoint conn savepoint)))))
```
I know what you're thinking. Why don't we just use Savepoints for everyone? H2's Savepoint impl seems broken and it errors out. So I implemented nested transactions this way for Postgres. H2 doesn't actually need it since the transaction doesn't become busted if you encounter an error inside it.

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,281 @@
             [metabase.util.i18n :refer [tru]]
-            [toucan.db :as db]))
-
-(defn- constraint->columns [conn constraint-name]
-  (->> ["select column_name from information_schema.constraint_column_usage where constraint_name = ?" constraint-name]
-       (jdbc/query conn)
-       (map :column_name)))
-
-(defn- violates-not-null-constraint [_conn error-message]
-  (let [[match? value column]
-        (re-find #"ERROR:\s+(\w+) value in column \"([^\"]+)\" violates not-null constraint" error-message)]
-    (when match?
-      [{:message (tru "{0} violates not-null constraint" value)
-        :column column}])))
-
-(defn- violates-unique-constraint [conn error-message]
-  (let [[match? constraint _value]
-        (re-find #"ERROR:\s+duplicate key value violates unique constraint \"([^\"]+)\"" error-message)]
-    (when match?
-      (let [columns (constraint->columns conn constraint)]
-        (mapv
-         (fn [column]
-           {:message (tru "violates unique constraint {0}" constraint)
-            :constraint constraint
-            :column column})
-         columns)))))
-
-(defn- update-or-delete-with-fk-constraint [conn error-message]
-  (let [[match? table constraint ref-table _columns _value]
-        (re-find #"ERROR:\s+update or delete on table \"([^\"]+)\" violates foreign key constraint \"([^\"]+)\" on table \"([^\"]+)\"" error-message)]
-    (when match?
-      (let [columns (constraint->columns conn constraint)]
-        (mapv
-         (fn [column]
-           {:message (tru "violates foreign key constraint {0}" constraint)
-            :table table
-            :ref-table ref-table
-            :constraint constraint
-            :column column})
-         columns)))))
```
Moved to `metabase.driver.postgres.actions` to as part of the fix for #23933

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,281 @@
-         violates-unique-constraint
-         update-or-delete-with-fk-constraint]))
+  {:arglists '([driver ^java.sql.Connection connection e]), :added "0.44.0"}
+  driver/dispatch-on-initialized-driver
+  :hierarchy #'driver/hierarchy)
```
Fix for #23934

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,281 @@
+  (let [default-method? (= (get-method parse-sql-error driver)
+                           (get-method parse-sql-error :default))]
+    (if default-method?
+      (ex-data e)
```
It's not a good idea to hardcode the assumption that `:postgres` has an implementation of `parse-sql-error` and no one else does. Instead of that check I tweaked it to check whether the current driver has a method for `parse-sql-error` that's different from the default method.

camsaul replied:
Oh yeah I should mention this but it's idiomatic to write `(if-not (= ...) ...)`, not `(if (not= ...) ...)`. 

https://guide.clojure.style/#if-not

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,281 @@
                     :status-code status-code}
-                   more-info))))
+                   more-info)
+            e)))
```
Fix for #23936

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,281 @@
-(defn- rollback! [tx]
-  (.rollback ^java.sql.Connection (:connection tx)))
-
-(def base-type->sql-type
-  "Mapping from base-types to postgres sql-types to e.g. be used for casting."
-  {:type/BigInteger          "BIGINT"
-   :type/Boolean             "BOOL"
-   :type/Date                "DATE"
-   :type/DateTime            "TIMESTAMP"
-   :type/DateTimeWithTZ      "TIMESTAMP WITH TIME ZONE"
-   :type/DateTimeWithLocalTZ "TIMESTAMP WITH TIME ZONE"
-   :type/Decimal             "DECIMAL"
-   :type/Float               "FLOAT"
-   :type/Integer             "INTEGER"
-   :type/IPAddress           "INET"
-   :type/Text                "TEXT"
-   :type/Time                "TIME"
-   :type/TimeWithTZ          "TIME WITH TIME ZONE"
-   :type/UUID                "UUID"})
```
Moved to `metabase.driver.postgres.actions` as part of the fix for #23933

camsaul committed:
Linter and test fixes :wrench:

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
+(defn- do-with-jdbc-transaction [database-id f]
+  (if *connection*
+    (f *connection*)
+    (let [jdbc-spec (sql-jdbc.conn/db->pooled-connection-spec database-id)]
```
I reworked the `row/*` Action impls to all use the new `with-jdbc-transaction` macro. It fixes a bunch of stuff including #23931

snoe replied:
> It fixes a bunch of stuff

Beyond the use of `db->pooled-connection-spec`. It would be useful to know why this is needed over/different than `jdbc/with-db-transaction` It also allows setting isolation levels and absorbs nested calls (which I believe is the reason for `*connection*` here) so a reimplementation is kinda worrying.

camsaul replied:
The main reason is while `clojure.java.jdbc` is good for doing simple stuff you do all the time, like fetching rows from the app DB or whatever, it does a lot of additional magic that we don't always want. When we want to do complicated fancy stuff like we're doing here, IMO working around the magic is more trouble than just hitting the JDBC API directly and doing exactly what we want. 

It's not really much more code, and it's easier to reason about because I can see what's happening right there instead of trying to remember the magic weird stuff `clojure.java.jdbc` may or may not be doing.

We need custom code to implement the custom nested transaction stuff anyway and since `clojure.java.jdbc` tries to prevent them using `jdbc/with-db-transaction` made the custom nested transaction code a bit more complicated than trying to code it against stuff that was already bespoke.

We're setting the best available transaction isolation level anyway using an existing driver method, so it's not like we would have been relying on `clojure.java.jdbc` to do that for us anyway

camsaul replied:
There are a few reasons:

1. `jdbc/with-db-transaction` "absorbs" nested transactions, but this only works if you take the transaction
connection spec bound there and pass it around explicitly to subsequent calls to `jdbc/with-db-transaction`.
This makes it hard to write bulk Actions as loops with repeated calls to single-row Actions. Of course, we could
just tweak the Actions so there was some way you could pass in an existing Connection or Connection spec for it
to use. But that would be a little more ugly and complicated than having a dynamic var. So we'd likely end up
with some sort of dynamic var and `with-` macro anyway to avoid repeated boilerplate. And it would likely make
`do-nested-transaction` harder to use or implement.

2. `jdbc/with-db-transaction` does a lot of magic that we don't necessarily want. Writing raw JDBC code is barely
any more cde and lets us have complete control over what happens and lets us see at a glance exactly what's
happening without having to keep `clojure.java.jdbc` magic in mind or work around it.

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
+             (ex-info (or (ex-message e) "Update action error.")
+                      (assoc e-data :status-code 400)))))))))
+
+(defn- select-created-row
+  "H2 and MySQL are dumb and `RETURN_GENERATED_KEYS` only returns the ID of the newly created row. This function will
+  `SELECT` the newly created row."
+  [driver raw-hsql conn result]
+  (let [select-hsql     (assoc raw-hsql
+                               :select [:*]
+                               ;; :and with a single clause will be optimized in HoneySQL
+                               :where (into [:and]
+                                            (for [[col val] result]
+                                              [:= (keyword col) val])))
+        select-sql-args (sql.qp/format-honeysql driver select-hsql)]
+    (log/tracef ":row/create SELECT HoneySQL:\n\n%s" (u/pprint-to-str select-hsql))
+    (log/tracef ":row/create SELECT SQL + args:\n\n%s" (u/pprint-to-str select-sql-args))
+    (first (jdbc/query {:connection conn} select-sql-args {:identifiers identity, :transaction? false}))))
```
I refactored the `:row/create` method a bit because it was kinda big (35 lines) and split some of it out into separate functions

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
-              (catch Exception e
-                (throw
-                 (ex-info "Create action error." (assoc (parse-error driver conn e) :status-code 400))))))
-       (try (let [col->val (jdbc/execute! conn (hformat/format create-hsql) {:return-keys true})]
```
You can see both

- #23932
- #23927

on this line; those are both fixed in the new version

camsaul commented on .clj-kondo/config.edn:
```diff
@@ -345,45 +345,49 @@
+ {metabase-enterprise.serialization.test-util/with-random-dump-dir      clojure.core/let
+  metabase.api.common/let-404                                           clojure.core/let
+  metabase.db.data-migrations/defmigration                              clojure.core/def
+  metabase.db.liquibase/with-liquibase                                  clojure.core/let
+  metabase.driver.mongo.query-processor/mongo-let                       clojure.core/let
+  metabase.driver.mongo.util/with-mongo-connection                      clojure.core/let
+  metabase.driver.sql-jdbc.actions/with-jdbc-transaction                clojure.core/let
+  metabase.driver.sql-jdbc.execute.diagnostic/capturing-diagnostic-info clojure.core/fn
+  metabase.integrations.ldap/with-ldap-connection                       clojure.core/fn
+  metabase.mbql.schema.macros/defclause                                 clj-kondo.lint-as/def-catch-all
+  metabase.models.setting.multi-setting/define-multi-setting            clojure.core/def
+  metabase.models.setting/defsetting                                    clj-kondo.lint-as/def-catch-all
+  metabase.public-settings.premium-features/defenterprise               clj-kondo.lint-as/def-catch-all
+  metabase.public-settings.premium-features/defenterprise-schema        clj-kondo.lint-as/def-catch-all
+  metabase.public-settings.premium-features/define-premium-feature      clojure.core/def
+  metabase.query-processor.error-type/deferror                          clojure.core/def
+  metabase.sync.util/sum-for                                            clojure.core/for
+  metabase.sync.util/with-emoji-progress-bar                            clojure.core/let
+  metabase.test.data.interface/defdataset                               clojure.core/def
+  metabase.test.data.interface/defdataset-edn                           clojure.core/def
+  metabase.test/are+                                                    clojure.test/are
+  metabase.test/defdataset                                              clojure.core/def
+  metabase.test/with-open-channels                                      clojure.core/let
+  metabase.test/with-temp-file                                          clojure.core/let
+  metabase.test/with-user-in-groups                                     clojure.core/let
+  metabase.util.files/with-open-path-to-resource                        clojure.core/let
+  metabase.util.ssh/with-ssh-tunnel                                     clojure.core/let
+
+  clojure.core.logic/defne           clj-kondo.lint-as/def-catch-all
+  clojurewerkz.quartzite.jobs/defjob clojure.core/defn
+  honeysql.helpers/defhelper         clj-kondo.lint-as/def-catch-all
+  honeysql.util/defalias             clojure.core/def
+  monger.operators/defoperator       clojure.core/def
+  potemkin.types/defprotocol+        clojure.core/defprotocol
+  potemkin.types/defrecord+          clojure.core/defrecord
+  potemkin.types/deftype+            clojure.core/deftype
+  potemkin/defprotocol+              clojure.core/defprotocol
+  potemkin/defrecord+                clojure.core/defrecord
+  potemkin/deftype+                  clojure.core/deftype
+  toucan.db/with-call-counting       clojure.core/fn}
```
I added `metabase.driver.sql-jdbc.actions/with-jdbc-transaction`, but these were out of order, so I didn't know where to add it. Also it looks like stuff was divided up into two groups -- Metabase stuff and 3rd-party stuff -- but `toucan.db/with-call-counting` was in the Metabase stuff group. Made my eyes burn. Fixed it and indented it nicely.

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
-        delete-hsql  (-> raw-hsql
-                         (dissoc :select)
-                         (assoc :delete []))]
-    (jdbc/with-db-transaction [tx (database->connection-spec driver database-id)]
```
- #23931

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
-
-(defmethod actions/perform-action!* [:sql-jdbc :row/update]
-  [driver _action _database {database-id :database :keys [update-row] :as query}]
-  (let [conn         (sql-jdbc.conn/db->pooled-connection-spec database-id)
```
- #23931

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
-(defmethod actions/perform-action!* [:sql-jdbc :row/update]
-  [driver _action _database {database-id :database :keys [update-row] :as query}]
-  (let [conn         (sql-jdbc.conn/db->pooled-connection-spec database-id)
-        raw-hsql     (qp.store/with-store
-                   (try
```
This indentation was busted

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
                          (assoc :update target-table
-                                :set (cast-values update-row (get-in query [:query :source-table]))))]
-    (try
-      (jdbc/with-db-transaction [tx (database->connection-spec driver database-id)]
```
- #23931

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
-                                :set (cast-values update-row (get-in query [:query :source-table]))))]
-    (try
-      (jdbc/with-db-transaction [tx (database->connection-spec driver database-id)]
-        (let [rows-updated (first (jdbc/execute! tx (hformat/format update-hsql)))]
```
- #23927
- #23932

camsaul commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
 
 (defmethod actions/perform-action!* [:sql-jdbc :row/create]
   [driver _action _database {database-id :database :keys [create-row] :as query}]
-  (let [conn        (sql-jdbc.conn/db->pooled-connection-spec database-id)
```
- #23931

camsaul commented on src/metabase/driver/sql_jdbc/connection.clj:
```diff
@@ -20,7 +20,11 @@
-  "Given a Database `details-map`, return a JDBC connection spec."
+  "Given a Database `details-map`, return a JDBC connection spec.
+
+  DO NOT USE THIS METHOD DIRECTLY UNLESS YOU KNOW WHAT YOU ARE DOING! THIS RETURNS AN UNPOOLED CONNECTION SPEC! IF YOU
+  WANT A CONNECTION SPEC FOR RUNNING QUERIES USE [[db->pooled-connection-spec]] INSTEAD WHICH WILL RETURN A *POOLED*
+  CONNECTION SPEC."
```
I wrote this note so hopefully

#23931

doesn't creep in again.

camsaul commented on test/metabase/actions/test_util.clj:
```diff
@@ -35,7 +35,7 @@
         (thunk))
       (finally
-        (let [{driver :engine, db-id :id} @db]
+        (when-let [{driver :engine, db-id :id} @db]
```
Minor fix -- if a test DB can't be created because it runs into an error, don't try to destroy the DB in the `finally` block. I ran into this when the `postgres` service on my computer shut down after running `apt upgrade`

camsaul commented on test/metabase/actions_test.clj:
```diff
@@ -0,0 +1,18 @@
+   [clojure.test :refer :all]
+   [metabase.actions :as actions]))
+
+(deftest normalize-as-mbql-query-test
+  (testing "Make sure normalize-as-mbql-query can exclude certain keys from normalization"
+    (is (= {:database    1
+            :type        :query
+            :updated-row {:my_snake_case_column 1000
+                          "CamelCaseColumn"     {:ABC 200}}
+            :query       {:source-table 2}}
+           (#'actions/normalize-as-mbql-query
+            {"database"   1
+             :updated_row {:my_snake_case_column 1000
+                           "CamelCaseColumn"     {:ABC 200}}
+             :query       {"source_table" 2}}
+            :exclude #{:updated-row})))))
```
Added test for the fix for 

- #23930

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -53,57 +53,75 @@
-    :request-body (assoc (mt/mbql-query categories {:filter [:= $id 10]})
-                         :update_row {:name "new-category-name"})
-    :expected     {:rows-updated [1]}}])
+(defn- format-field-name
+  "Format `field-name` appropriately for the current driver (e.g. uppercase it if we're testing against H2)."
+  [field-name]
+  (keyword (mt/format-name (name field-name))))
```
Needed because of the fixes for 

- #23927
- #23930
- #23932

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -53,57 +53,75 @@
+(defn- categories-row-count []
+  (first (mt/first-row (mt/run-mbql-query categories {:aggregation [[:count]]}))))
+
+(deftest create-test
```
I split out/replaced `happy-path-test` and `create-update-delete-test` with separate `create-test`, `update-test`, and `delete-test`s. They're testing the same thing but it's a lot easier to read them now.

camsaul replied:
It also makes test failures easier to debug since they're more granular -- a bug in `create` won't cause `update` to fail

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -53,57 +53,75 @@
-      (let [[create update delete] (mock-requests)
-            {created-id :id :as created-row}
-            (:created-row (mt/user-http-request :crowberto :post 200 (:action create) (:request-body create)))]
-        (is (= [:id :name] (keys created-row))
-            "Create should return the entire row")
-        (is (= "created_row" (:name created-row))
-            "Create should return the correct value for name")
```
`schema=` is perfectly suited for this type of thing and a lot easier to read, especially if the test fails

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -53,57 +53,75 @@
-        (is (= (:expected update) (mt/user-http-request :crowberto :post 200 (:action update)
-                                                        (assoc (mt/mbql-query categories {:filter [:= $id created-id]})
-                                                               :update_row {:name "updated_row"})))
+(deftest create-invalid-data-test
+  (testing "POST /api/action/row/create -- invalid data"
+    (mt/test-drivers (mt/normal-drivers-with-feature :actions)
+      (actions.test-util/with-actions-test-data-and-actions-enabled
+        (is (= 75
+               (categories-row-count)))
+        (is (schema= {:message  (s/constrained
+                                 s/Str
+                                 (case driver/*driver*
+                                   :h2       #(str/starts-with? % "Data conversion error converting \"created_row\"")
+                                   :postgres #(str/starts-with? % "ERROR: invalid input syntax for type integer: \"created_row\"")))
+                      s/Keyword s/Any}
+                     ;; bad data -- ID is a string instead of an Integer.
+                     (mt/user-http-request :crowberto :post 400
+                                           "action/row/create"
+                                           (assoc (mt/mbql-query categories) :create-row {(format-field-name :id) "created_row"}))))
+        (testing "no row should have been inserted"
+          (is (= 75
+                 (categories-row-count))))))))
```
Added a test for `row/create` with invalid data.

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -114,6 +132,29 @@
           (let [request-body (mt/mbql-query categories {:filter [:= $id 22]})]
             (mt/user-http-request :crowberto :post 400 "action/row/delete" request-body))))))
 
+(defn- mock-requests
+  "Mock requests for testing validation for various actions. Don't use these for happy path tests! It's way too hard to
+  wrap your head around them. Use them for validating preconditions and stuff like that."
+  []
+  [{:action       "action/row/create"
+    :request-body (assoc (mt/mbql-query categories) :create-row {(format-field-name :name) "created_row"})
+    :expect-fn    (fn [result]
+                    ;; check that we return the entire row
+                    (is (schema= {:created-row {(format-field-name :id)   su/IntGreaterThanZero
+                                                (format-field-name :name) su/NonBlankString}}
+                                 result)))}
+   {:action       "action/row/update"
+    :request-body (assoc (mt/mbql-query categories {:filter [:= $id 1]})
+                         :update_row {(format-field-name :name) "updated_row"})
+    :expected     {:rows-updated [1]}}
+   {:action       "action/row/delete"
+    :request-body (mt/mbql-query categories {:filter [:= $id 1]})
+    :expected     {:rows-deleted [1]}}
+   {:action       "action/row/update"
+    :request-body (assoc (mt/mbql-query categories {:filter [:= $id 10]})
+                         :update_row {(format-field-name :name) "new-category-name"})
+    :expected     {:rows-updated [1]}}])
```
This is untouched but I moved it down in the namespace a bit to try to make it clear it's not being used for the happy path `row/*` action tests anymore.

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -168,28 +209,34 @@
-                 (:message (mt/user-http-request :crowberto :post 400 action query-that-returns-more-than-one))))
-        (is (= result-count (count (mt/rows (qp/process-query query-that-returns-more-than-one))))
-            "The result-count after a rollback must remain the same!")))))
+  (mt/test-drivers (mt/normal-drivers-with-feature :actions)
```
I reworked a lot of the tests to add `mt/test-drivers (mt/normal-drivers-with-feature :actions)` so they run against Postgres as well. Previously, they only ran against H2

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -168,28 +209,34 @@
+  (mt/test-drivers (mt/normal-drivers-with-feature :actions)
+    (actions.test-util/with-actions-enabled
+      (let [query-that-returns-more-than-one (assoc (mt/mbql-query users {:filter [:>= $id 1]})
+                                                    :update_row {(format-field-name :name) "new-name"})
```
We have to use  `format-field-name` in a bunch of places now because the column is called `"NAME"` for H2, not `"name"`. Fixing 

- #23927

means we have to pass in the correct case now. But that's ok because doing so fixes

- #23930
- #23932

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -168,28 +209,34 @@
+      (let [query-that-returns-more-than-one (assoc (mt/mbql-query users {:filter [:>= $id 1]})
+                                                    :update_row {(format-field-name :name) "new-name"})
+            result-count                     (count (mt/rows (qp/process-query query-that-returns-more-than-one)))]
+        (is (< 1 result-count))
+        (doseq [{:keys [action]} (filter #(= "action/row/update" (:action %)) (mock-requests))
+                :when            (not= action "action/row/create")] ;; the query in create is not used to select values to act upopn.
+          (is (schema= {:message #"Sorry, this would update [\d|,]+ rows, but you can only act on 1"
+                        s/Keyword s/Any}
+                       (mt/user-http-request :crowberto :post 400 action query-that-returns-more-than-one)))
+          (is (= result-count (count (mt/rows (qp/process-query query-that-returns-more-than-one))))
+              "The result-count after a rollback must remain the same!"))))))
```
The rest of this stuff is mostly unchanged but I replaced

```clj
(re= <pattern> (:message <results>))
```

with 

```clj
(schema= {:message <pattern>, s/Keyword s/Any} <results>
```

because with the old form the test failure is meaningless if the response does not contain a `:message` at all. With the new form you see the entire response when it fails

camsaul commented on test/metabase/api/database_test.clj:
```diff
@@ -299,7 +299,7 @@
                                                              :has_field_values  "list"
                                                              :database_position 1
-                                                             :database_required false})]
+                                                             :database_required true})]
```
Now that `test-data.categories.name` is `NOT NULL` I had to update two metadata tests.

camsaul commented on test/metabase/test/data/sql.clj:
```diff
@@ -198,38 +198,43 @@
   tx/dispatch-on-driver-with-test-extensions
   :hierarchy #'driver/hierarchy)
 
+(defn- format-and-quote-field-name [driver field-name]
+  (sql.u/quote-name driver :field (ddl.i/format-name driver field-name)))
+
+(defn- field-definition-sql
+  [driver {:keys [field-name base-type field-comment not-null?], :as field-definition}]
+  (let [field-name (format-and-quote-field-name driver field-name)
+        field-type (or (cond
+                         (and (map? base-type) (contains? base-type :native))
+                         (:native base-type)
+
+                         (and (map? base-type) (contains? base-type :natives))
+                         (get-in base-type [:natives driver])
+
+                         base-type
+                         (field-base-type->sql-type driver base-type))
+                       (throw (ex-info (format "Missing datatype for field %s for driver: %s"
+                                               field-name driver)
+                                       {:field  field-definition
+                                        :driver driver})))
+        not-null       (when not-null?
+                         "NOT NULL")
+        inline-comment (inline-column-comment-sql driver field-comment)]
+    (str/join " " (filter some? [field-name field-type not-null inline-comment]))))
```
I refactored this a bit and moved the code for field definition DDS into its own function. The `create-table-sql` method was kind of big at 30 lines; moving half of that out into a separate function makes both easier to read.

I also refactored this stuff a bit to make the way the string gets built from different parts more obvious.

I also added support for `NOT NULL` as mentioned earlier, needed to be able to easily test failing inserts

camsaul commented on test/metabase/api/action_test.clj:
```diff
@@ -205,3 +252,56 @@
+          (is (= 75
+                 (categories-row-count)))
+          (testing "Should report indices of bad rows"
+            (is (schema= {:errors [(s/one {:index (s/eq 1)
+                                           :error (s/constrained
+                                                   s/Str
+                                                   (case driver/*driver*
+                                                     :h2       #(str/starts-with? % "NULL not allowed for column \"NAME\"")
+                                                     :postgres #(str/starts-with? % "ERROR: null value in column \"name\"")))}
```
This error shape isn't exactly what we wanted in the tech design doc but seemed close enough for an MVP. We have the row `:index` like we wanted but column errors are not keyed by the column name -- it seemed hard to do and this PR had enough going on already. We can add that stuff as a follow-on I think.

camsaul commented on test/metabase/api/table_test.clj:
```diff
@@ -435,19 +435,19 @@
-                              :base_type        "type/BigInteger"
-                              :effective_type   "type/BigInteger"
-                              :has_field_values "none"
+                              :name              "ID"
+                              :display_name      "ID"
+                              :database_type     "BIGINT"
+                              :base_type         "type/BigInteger"
+                              :effective_type    "type/BigInteger"
+                              :has_field_values  "none"
```
indentation fix

camsaul committed:
Fix bad metadata (thanks Eastwood!)

snoe commented on src/metabase/driver/sql_jdbc/actions.clj:
```diff
@@ -1,235 +1,277 @@
+(defmethod actions/perform-action!* [:sql-jdbc :bulk/create]
+  [driver _action {database-id :id, :as database} {:keys [table-id rows]}]
+  (log/tracef "Inserting %d rows" (count rows))
+  (with-jdbc-transaction [conn database-id]
+    (transduce
+     (map-indexed (fn [i row] [i row]))
+     (fn
+       ([]
+        [[] []])
+       ([[errors created-rows]]
+        ;; if there were any errors throw an Exception with all the error messages.
+        (when (seq errors)
+          (.rollback conn)
+          (throw (ex-info (tru "Error(s) inserting rows.")
+                          {:status-code 400, :errors errors})))
+        ;; if there we no errors then return the created rows.
+        {:created-rows created-rows})
+       ([[errors created-rows] [row-index row]]
+        (try
+          (let [result (do-nested-transaction
+                        driver
+                        conn
+                        (fn []
+                          (actions/perform-action!*
+                           driver
+                           :row/create
+                           database
+                           {:database   database-id
+                            :type       :query
+                            :query      {:source-table table-id}
+                            :create-row row})))]
+            [errors (conj created-rows (:created-row result))])
+          (catch Throwable e
+            [(conj errors {:index row-index, :error (ex-message e)})
+             created-rows]))))
+     rows)))
```
It's not too clear why nested transactions are necessary.

I believe it's because: if row #1 fails, normally during a transaction then subsequent rows would fail with a "transaction rolled back" error even if those rows would normally succeed. This allows collecting "real" errors and then rolling back the whole thing at the end.

Maybe a comment explaining that would help.

snoe reviewed (changes requested):
Looks good, just a bit confused about some of the transaction details.

bshepherdson committed:
Serdes v2 for Native Query Snippets (#23961)

Also standardizes date/time output to `ZonedDateTime`, rather than whatever the JDBC happens to return.

camsaul committed:
Merge branch 'pod-writeback' into bulk-insert-action

camsaul committed:
Add some comments about why we're doing what we're doing

camsaul commented:
@snoe I wrote comments adding more context about the transactions stuff.
